File: aurastamp-npm/src/aurastamp-image.ts
import axios, { AxiosInstance } from 'axios';
import FormData from 'form-data';
import { readFileSync, writeFileSync } from 'fs';
import mime from 'mime-types';
import path from 'path';

type ReturnType = 'base64' | 'buffer';

type ModelKind = 'the';

type EncodeResponse = {
  data: string;
};

type DecodeResponse = {
  dislike_cnt: number;
  hash_string: string;
  like_cnt: number;
  secret: string;
  secret_image: null | string;
  view_cnt: number;
};

const aurastampApiUrl = (process.env.AURASTAMP_API_URL = 'https://api.aurastamp.com');

export class AurastampImage {
  private readonly client: AxiosInstance;

  constructor({ serverUrl }: { serverUrl?: string }) {
    this.client = axios.create({
      baseURL: serverUrl ?? aurastampApiUrl,
    });
  }

  async encode({
    model = 'the',
    image,
    message,
    hiddenImage,
    returnType = 'base64',
    contentType,
    filename,
    headers = {},
  }: {
    model: ModelKind;
    image: Buffer;
    message: string;
    hiddenImage?: Buffer;
    returnType?: ReturnType;
    contentType?: string;
    filename: string;
    headers?: Record<string, any>;
  }) {
    const formData = new FormData();
    formData.append('file', image, { filename, contentType });
    formData.append('model_name', model);
    formData.append('text', message);
    formData.append('return_type', returnType);
    if (hiddenImage) {
      formData.append('media', hiddenImage);
    }

    const response = await this.client.post('/encode', formData, {
      headers: {
        ...formData.getHeaders(),
        ...headers,
      },
    });

    return response;
  }

  async encodeLocalFile({
    model = 'the',
    message,
    returnType = 'base64',
    imageFilePath,
    hiddenImageFilePath,
    outputPath,
    headers = {},
  }: {
    model: ModelKind;
    message: string;
    returnType?: ReturnType;
    imageFilePath: string;
    hiddenImageFilePath?: string;
    outputPath?: string;
    headers?: Record<string, any>;
  }) {
    const formData = new FormData();

    if (!imageFilePath.startsWith('/')) {
      // if relative path, convert it to absolute path
      imageFilePath = path.join(process.cwd(), imageFilePath);
    }

    const image = readFileSync(imageFilePath);
    const contentType = mime.contentType(path.extname(imageFilePath)) || undefined;
    formData.append('file', image, {
      filename: path.basename(imageFilePath),
      contentType,
    });
    formData.append('model_name', model);
    formData.append('text', message);
    formData.append('return_type', returnType);
    if (hiddenImageFilePath) {
      const hiddenImage = readFileSync(hiddenImageFilePath);
      formData.append('media', hiddenImage);
    }

    const response = await this.client.post<never, EncodeResponse>('/encode', formData, {
      headers: {
        ...formData.getHeaders(),
        responseType: 'text',
        // responseType: 'arraybuffer',
        responseEncoding: 'base64',
        ...headers,
      },
    });

    if (outputPath) {
      writeFileSync(outputPath, response.data, { encoding: 'base64' });
    }

    return response;
  }

  async decode({
    model,
    image,
    filename,
    contentType,
    headers = {},
  }: {
    model: ModelKind;
    image: Buffer;
    filename: string;
    contentType?: string;
    headers?: Record<string, any>;
  }) {
    const formData = new FormData();
    formData.append('file', image, { filename, contentType });
    formData.append('model_name', model);
    return this.client.post<DecodeResponse>('/decode', formData, {
      headers: { ...formData.getHeaders(), ...headers },
    });
  }

  async decodeLocalFile({
    model,
    imageFilePath,
    outputPath,
    headers = {},
  }: {
    model: ModelKind;
    imageFilePath: string;
    outputPath?: string;
    headers?: Record<string, any>;
  }) {
    // read file and decode it
    throw new Error('not implemented');
  }
}

File: aurastamp-pip/aurastamp/aurastamp_image.py
import requests
import io
import shutil
import base64
from PIL import Image
from requests_toolbelt.multipart.encoder import MultipartEncoder


class AurastampImage(object):
    def __init__(self, server_url):
        default_server_url = 'https://api.aurastamp.com'
        if not server_url:
            server_url = default_server_url
        
        self.server_url = server_url

    def encode(self,
               image,
               message,
               hidden_image=None,
               model='the',
               return_type='base64',
               headers={}):
        
        field_data = {}
        field_data['file'] = ("filename", image)
        field_data['model_name'] = model
        field_data['message'] = message
        field_data['return_type'] = return_type
        if hidden_image:
            field_data['media'] = hidden_image

        m = MultipartEncoder(fields=field_data)
        r = requests.post(
            self.server_url+"/encode",
            data=m,
            headers={"Content-Type": m.content_type, **headers}
        )

        return r    

    def encode_local_file(self,
                          image_file_path,
                          message,
                          hidden_image_file_path=None,
                          output_path=None,
                          model='the',
                          return_type='base64',
                          headers={}):
                    
        field_data = {}
        image = Image.open(image_file_path)
        image_bytes_io = io.BytesIO()
        image.save(image_bytes_io, image.format.lower())

        # image
        field_data['file'] = (image.filename, image_bytes_io)
        field_data['model_name'] = model
        field_data['message'] = message
        field_data['return_type'] = return_type
        if hidden_image_file_path:
            hidden_image = Image.open(hidden_image_file_path)
            field_data['media'] = hidden_image

        m = MultipartEncoder(fields=field_data)
        r = requests.post(
            self.server_url+"/encode",
            data=m,
            headers={"Content-Type": m.content_type,
                     **headers}
        )

        if output_path:
              with open(output_path, 'wb') as f:
                container = base64.b64decode(r.content)
                container_image = io.BytesIO(container)
                shutil.copyfileobj(container_image, f)         

        return r


    def decode(self,
               image,
               model='the',
               headers={}):
       
        field_data = {}
        field_data['file'] = ("filename", image)
        field_data['model_name'] = model

        m = MultipartEncoder(fields=field_data)
        r = requests.post(
            self.server_url+"/decode",
            data=m,
            headers={"Content-Type": m.content_type, **headers}
        )

        return r.json()

    def decode_local_file(self,
                          image_file_path,
                          model='the',
                          output_path=None,
                          headers={}):

        field_data = {}
        image = Image.open(image_file_path)
        image_bytes_io = io.BytesIO()
        image.save(image_bytes_io, image.format.lower())

        # image
        field_data['file'] = (image.filename, image_bytes_io)
        field_data['model_name'] = model

        m = MultipartEncoder(fields=field_data)
        r = requests.post(
            self.server_url+"/decode",
            data=m,
            headers={"Content-Type": m.content_type,

                     **headers}
        )

        r = r.json()

        if output_path and r['secret_image']:
            secret_r = requests.get(r['secret_image'], stream=True)
            if secret_r.status_code == 200:
                with open(output_path, 'wb') as f:
                    secret_r.raw.decode_content = True
                    shutil.copyfileobj(secret_r.raw, f)         

        return r
